pkg/util/stream/publisher.go

```diff
@@ -0,0 +1,154 @@
+			select {
+			case s.values <- val:
+				// This send might block if the subscriber has quit receiving from its
+				// values channel.
+			case <-s.canceled:
+			}
+		}
+	}
+	// Send the close to all subscribers
+	p.mutex.Lock()
+	subs := p.subscribers
+	p.in = nil
+	p.subscribers = nil
+	p.closed = true
+	p.mutex.Unlock()
+	for _, s := range subs {
+		s.waitForReady()
+		s.close()
+	}
+}
+
+// Subscribe creates a new subscription to the publisher's data stream. If a nil channel
+// argument is passed, one will be created for the caller.
+func (p *StringValuePublisher) Subscribe(values chan string) *StringSubscription {
+	if values == nil {
+		values = make(chan string)
+	}
+	s := &StringSubscription{
+		values:      values,
+		initialized: make(chan struct{}),
+		canceled:    make(chan struct{}),
+		publisher:   p,
+	}
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	// Send the first value in another goroutine to avoid blocking this one
+	if p.closed {
+		go func(firstValue string) {
+			s.sendFirst(firstValue)
+			s.close()
+		}(p.value)
+	} else {
+		// Append doesn't modify existing slices, so it's still safe.
+		p.subscribers = append(p.subscribers, s)
+		go s.sendFirst(p.value)
+	}
+	return s
+}
+
+func (p *StringValuePublisher) unsubscribe(subscription *StringSubscription) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	for i, s := range p.subscribers {
+		if s == subscription {
+			// Subscribe() and read() assume their slices' contents are immutable
+			subLen := len(p.subscribers)
+			newSub := make([]*StringSubscription, subLen-1)
+			copy(newSub[:i], p.subscribers[:i])
+			copy(newSub[i:subLen-1], p.subscribers[i+1:subLen])
+			p.subscribers = newSub
+			return
+		}
+	}
+}
+
+// Chan returns a channel that delivers string values from the publisher. Returns a
+// receive-only channel with a buffer size of 1.
+func (s *StringSubscription) Chan() <-chan string {
+	return s.values
+}
+
+// Unsubscribe cancels the subscription. Sends to the publisher's input channel that
+// happen after this method returns will not be delivered to this subscription's channel;
+// earlier sends are concurrent and might be delivered.
+func (s *StringSubscription) Unsubscribe() {
+	s.publisher.unsubscribe(s)
+	s.publisher = nil
+	close(s.canceled)
+}
+
+// waitForReady blocks until either (1) the subscription has received its first value and
+// is ready to receive further streaming values, or (2) the subscription is canceled.
+func (s *StringSubscription) waitForReady() {
+	select {
+	case <-s.initialized:
+	case <-s.canceled:
+	}
+}
+
+// sendFirst sends the first value to the subscription stream, then marks the subscription
+// as ready for more streaming data.
+func (s *StringSubscription) sendFirst(val string) {
+	select {
+	case s.values <- val:
+		close(s.initialized)
+	case <-s.canceled:
+	}
+}
+
+// close the subscription's data stream if the subscription wasn't previously canceled
+func (s *StringSubscription) close() {
+	select {
+	default:
```

Comment:
The main job of the method is to close its output socket, so I wrote that first. It's also makes the method similar to `waitForReady()` and `sendFirst()` in that checking for cancellation is always written last.